### PR TITLE
Additional Mathematical Opcodes

### PR DESCRIPTION
**Pull request by amityadav0**
- Opcodes: divmodw, exp, expw, shr, shl, sqrt

### CODE REVIEW

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 1]
```diff
@@ -2789,3 +2789,216 @@ export class ByteAdd extends Op {
...
+
+    stack.push(high);
+    stack.push(low);
+
+    const remainder = dividend % divisor;
+    low = remainder & MAX_UINT64;
+    this.checkOverflow(low, this.line);
+
+    high = remainder >> BigInt('64');
+    this.checkOverflow(high, this.line);
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:2844]: Could you please add a reference link?

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 2]
```diff
@@ -2789,3 +2789,216 @@ export class ByteAdd extends Op {
...
+
+    stack.push(high);
+    stack.push(low);
+  }
+}
+
+// A raised to the Bth power. Panic if A == B == 0 and on overflow
+// Pops: ... stack, {uint64 A}, {uint64 B}
+// Pushes: uint64
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:2853]: `uint64` or `uint128`?

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 3]
```diff
@@ -2789,3 +2789,216 @@ export class ByteAdd extends Op {
...
+      x1 = BigInt(Math.floor(Math.sqrt(Number(value)))) - 3n;
+    } else {
+      x1 = (1n << 52n) - 2n;
+    }
+
+    let x0 = -1n;
+    while ((x0 !== x1 && x0 !== (x1 - 1n))) {
+      x0 = x1;
+      x1 = ((value / x0) + x0) >> 1n;
+    }
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:3000]: could you add a link to reference?

File: packages/runtime/src/interpreter/opcode.ts [hunk 4]
```diff
@@ -28,6 +28,17 @@ export class Op {
...
+  /**
+   * asserts number is less than or equal to MAX_UINT128
+   * @param num number to check
+   * @param line line number in TEAL file
+   */
+  checkOverflowU128 (num: bigint, line: number): void {
+    if (num > MAX_UINT128) {
+      throw new RuntimeError(RUNTIME_ERRORS.TEAL.UINT128_OVERFLOW, { line: line });
+    }
+  }
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode.ts:40]: Why not use `checkOverflow` function above? We could generalize  it as `checkUintOverflow` & pass `MAX_UINT64`/`MAX_UINT128`. Errors could be generalized as well.

File: packages/runtime/src/lib/constants.ts [hunk 5]
```diff
@@ -2,6 +2,7 @@ import { EncodedAssetParams, EncodedLocalStateSchema, EncodedTransaction } from
 
 export const MIN_UINT64 = 0n;
 export const MAX_UINT64 = 0xFFFFFFFFFFFFFFFFn;
+export const MAX_UINT128 = (2n ** 128n) - 1n;
```
Comment (ratik21) [inline packages/runtime/src/lib/constants.ts:5]: Shouldn't we use a fixed(constant) value? Won't this be computed each time when we run `checkOverflowUint128`

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 6]
```diff
@@ -4376,4 +4377,176 @@ describe("Teal Opcodes", function () {
...
+      );
+
+      stack.push(2n);
+      stack.push(66n);
+
+      expectRuntimeError(
+        () => op.execute(stack),
+        RUNTIME_ERRORS.TEAL.UINT64_OVERFLOW
+      );
+    });
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:4458]: Could you please add a test for max value ( ~ MAX_UINT64)?

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 7]
```diff
@@ -4376,4 +4377,176 @@ describe("Teal Opcodes", function () {
...
+      stack.push(0n);
+      stack.push(0n);
+
+      expectRuntimeError(
+        () => op.execute(stack),
+        RUNTIME_ERRORS.TEAL.EXP_ERROR
+      );
+
+      stack.push(2n);
+      stack.push(128n);
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:4488]: Same, could you please add a test where we construct a 128bit (64 + 64) output here?

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 8]
```diff
@@ -4376,4 +4377,176 @@ describe("Teal Opcodes", function () {
...
+
+      assert.equal(stack.pop(), exp);
+
+      stack.push(5n);
+      stack.push(21n);
+
+      exp = 5n >> 21n;
+      op.execute(stack);
+
+      assert.equal(stack.pop(), exp);
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:4531]:
```suggestion
      op.execute(stack);
      assert.equal(stack.pop(), exp);

      stack.push(5n);
      stack.push(21n);
      exp = 5n >> 21n;
      op.execute(stack);
      assert.equal(stack.pop(), exp);
```

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 1 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode-list.ts:2844]: yes

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 2 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode-list.ts:2853]: uint64

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 3 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode-list.ts:3000]: added at 2978

File: packages/runtime/src/interpreter/opcode.ts (hunk 4 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode.ts:40]: ok

File: packages/runtime/src/lib/constants.ts (hunk 5 above)
Comment (amityadav0) [reply packages/runtime/src/lib/constants.ts:5]: yes

File: packages/runtime/test/src/interpreter/opcode-list.ts (hunk 7 above)
Comment (amityadav0) [reply packages/runtime/test/src/interpreter/opcode-list.ts:4488]: it is present at 4994 where 2^66 is a (64 + 64) output

File: packages/runtime/test/src/interpreter/opcode-list.ts (hunk 6 above)
Comment (amityadav0) [reply packages/runtime/test/src/interpreter/opcode-list.ts:4458]: yes

**Review by ratik21 — APPROVED** [review]

**Pull request merged**